### PR TITLE
Add Cylinder jwt to Scabbard cli

### PR DESCRIPTION
**Pull request by isabeltomb**
- Create `ScabbardClientBuilder` for building scabbard clients. Has a `with_url` function to set the url field and a `with_auth` function to set the auth field. `with_auth` is behind the experimental feature `client-jwt`.

- Modifies client functions to submit the string stored in the `auth` field in an authorization header when the `client-jwt` feature is enabled and requests are made to the REST API.

- Add a `create_cylinder_jwt_auth` function to create the auth string that is submitted to the REST API in an "Authorization" header. The auth string contains a json web token created and signed using cylinder jwt. The function is behind an experimental feature scabbard-cli-jwt.

- Add a `--key` option to the scabbard cli subcommands that don't already have it.

- Use ScabbardClientBuilder to create clients in scabbard cli, when the scabbard-cli-jwt feature is enabled create an auth string containing a json web token and call the builder's `with_auth` function.

**Testing:** 

run gameroom (or other node + smart contract setup) so scabbard cli can be used to interact with a smart contract
`docker-compose -f examples/gameroom/docker-compose.yaml up --build`

Set up a circuit in gameroom

from the splinter/cli directory get the circuit ID of the game room circuit
`cargo run --features splinter-cli-jwt circuit list --url http://localhost:8088`

from the splinter/cli directory get the service ID of the acme node
`cargo run --features splinter-cli-jwt circuit show --url http://localhost:8088 <circuit ID>`

from the scabbard/cli directory run a scabbard command with a valid key name (or no key name to use default key in home directory)
` cargo run --features scabbard-cli-jwt contract list --url http://localhost:8088 --service-id <circuit_ID>::<service_ID> --key <key_name>`

verify command was successful and the xo contract is printed in table format

run a scabbard command with an invalid key name
` cargo run --features scabbard-cli-jwt contract list --url http://localhost:8088 --service-id <circuit_ID>::<service_ID> --key bad_key`

verify command fails and gives an error message stating the key could not be found

### CODE REVIEW

File: services/scabbard/libscabbard/src/client/mod.rs [hunk 1]
```diff
@@ -68,7 +114,15 @@ impl ScabbardClient {
...
         debug!("Submitting batches via {}", url);
-        let request = Client::new().post(url).body(body);
+        // Allowing unused_mut because request must be mutable if experimental feature
+        // client-jwt is enabled, if feature is removed unused_mut notation can be removed
+        #[allow(unused_mut)]
+        let mut request = Client::new().post(url).body(body);
+
+        #[cfg(feature = "client-jwt")]
+        {
+            request = request.header("Authorizaiton", &self.auth);
```
Comment (shannynalayna) [inline services/scabbard/libscabbard/src/client/mod.rs:124]: s/"Authorizaiton"/"Authorization", also throughout this file

**Review by shannynalayna — CHANGES_REQUESTED** [review]

File: services/scabbard/libscabbard/src/client/mod.rs (hunk 1 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/client/mod.rs]: Fixed

File: services/scabbard/libscabbard/src/client/mod.rs [hunk 2]
```diff
@@ -29,18 +29,64 @@ use super::protocol::SCABBARD_PROTOCOL_VERSION;
...
+    auth: Option<String>,
+}
+
+impl ScabbardClientBuilder {
+    /// Creates a new ScabbardClientBuilder
+    pub fn new() -> Self {
+        ScabbardClientBuilder::default()
+    }
+
+    /// Sets the url
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/client/mod.rs]: Added more detail to the comments and fixed punctuation

File: services/scabbard/cli/src/key.rs [hunk 3]
```diff
@@ -85,6 +87,37 @@ fn determine_key_file_path(key: &str) -> Result<PathBuf, CliError> {
...
+
+    let private_key = load_user_key(key_name, default_path_string).map_err(|err| {
+        CliError::action_error_with_source("Unable to get private key from file", err.into())
+    })?;
+
+    let context = Secp256k1Context::new();
+    let signer = context.new_signer(private_key);
+
+    let encoded_token = JsonWebTokenBuilder::new().build(&*signer).map_err(|err| {
+        CliError::action_error_with_source("failed to build json web token", err.into())
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply services/scabbard/cli/src/key.rs:121]: fixed

File: services/scabbard/cli/src/key.rs [hunk 4]
```diff
@@ -85,6 +87,37 @@ fn determine_key_file_path(key: &str) -> Result<PathBuf, CliError> {
     }
 }
 
+#[cfg(feature = "scabbard-cli-jwt")]
+// build an authorization string containing a signed json web token
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply services/scabbard/cli/src/key.rs]: Added a detailed doc comment

**Review by shannynalayna — APPROVED** [review]

File: services/scabbard/libscabbard/src/client/mod.rs [hunk 5]
```diff
@@ -29,18 +29,69 @@ use super::protocol::SCABBARD_PROTOCOL_VERSION;
...
+        self.url = Some(url.into());
+        self
+    }
+    /// Sets the `auth` field of the ScabbardClientBuilder. The `auth` string contains a
+    /// signed json web token and will be passed to the Splinter REST API in an
+    /// Authorization header.
+    ///
+    /// # Arguments
+    ///
+    /// * `auth` - The authorization string containing a signed json web token.
```
Comment (ltseeley) [inline services/scabbard/libscabbard/src/client/mod.rs:62]: The auth field isn't necessarily going to be a JWT; it could be any other auth header supported by splinterd. It would be sufficient to just say that the value given here will be provided in the `Authorization` header of REST API requests.

File: services/scabbard/libscabbard/src/client/mod.rs [hunk 6]
```diff
@@ -29,18 +29,69 @@ use super::protocol::SCABBARD_PROTOCOL_VERSION;
...
+    /// as the bind endpoint for the Splinter REST API.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - The URL of the bind endpoint of the Splinter REST API.
+    pub fn with_url(mut self, url: &str) -> Self {
+        self.url = Some(url.into());
+        self
+    }
+    /// Sets the `auth` field of the ScabbardClientBuilder. The `auth` string contains a
```
Comment (ltseeley) [inline services/scabbard/libscabbard/src/client/mod.rs:56]: Should have a blank line between method definitions.

File: services/scabbard/libscabbard/src/client/mod.rs [hunk 7]
```diff
@@ -29,18 +29,69 @@ use super::protocol::SCABBARD_PROTOCOL_VERSION;
 
 pub use error::ScabbardClientError;
 
+/// Builder for building a ScabbardClient.
```
Comment (ltseeley) [inline services/scabbard/libscabbard/src/client/mod.rs:32]: In doc comments, it's general best to put names of code items in backticks so they are formatted as code. When referencing an item like this, it's even better to add a link like this: ``[`ScabbardClient`](crate::client::ScabbardClient)``.

File: services/scabbard/libscabbard/src/client/mod.rs [hunk 8]
```diff
@@ -29,18 +29,69 @@ use super::protocol::SCABBARD_PROTOCOL_VERSION;
...
+#[derive(Default)]
+pub struct ScabbardClientBuilder {
```
Comment (ltseeley) [inline services/scabbard/libscabbard/src/client/mod.rs:34]: Generally we define builders in a `builder` submodule and `pub use` them at the level of the thing they construct. This keeps this higher level module less cluttered.

File: services/scabbard/libscabbard/src/client/mod.rs [hunk 9]
```diff
@@ -29,18 +29,69 @@ use super::protocol::SCABBARD_PROTOCOL_VERSION;
...
+                ScabbardClientError::new("Failed to build client, jwt authorization not provided")
+            })?,
+        })
+    }
+}
+
 /// A client that can be used to interact with scabbard services on a Splinter node.
 pub struct ScabbardClient {
     url: String,
+    #[cfg(feature = "client-auth")]
```
Comment (ltseeley) [inline services/scabbard/libscabbard/src/client/mod.rs:90]: This change should be in a separate commit, since it's distinct from the addition of a builder.

File: services/scabbard/cli/src/main.rs [hunk 10]
```diff
@@ -150,6 +150,11 @@ fn run() -> Result<(), CliError> {
                                 .takes_value(true)
                                 .possible_values(&["human", "csv"])
                                 .default_value("human"),
+                            Arg::with_name("key")
```
Comment (ltseeley) [inline services/scabbard/cli/src/main.rs:155]: We should probably use a different name than `key` here. The `--key` argument is used by some commands to specify the key for signing transactions; it's very possible that that's a different key than the one used to authenticate with the REST API.

**Review by ltseeley — CHANGES_REQUESTED** [review]

File: services/scabbard/libscabbard/src/client/mod.rs (hunk 6 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/client/mod.rs]: Fixed

File: services/scabbard/libscabbard/src/client/mod.rs (hunk 5 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/client/mod.rs]: Fixed comment to only reference auth string and not jwt

File: services/scabbard/libscabbard/src/client/mod.rs (hunk 7 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/client/mod.rs]: Fixed

File: services/scabbard/libscabbard/src/client/mod.rs (hunk 8 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/client/mod.rs]: Made a builder submodule

File: services/scabbard/libscabbard/src/client/mod.rs (hunk 9 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/client/mod.rs:37]: Moved this to its own commit

File: services/scabbard/cli/src/main.rs (hunk 10 above)
Comment (isabeltomb) [reply services/scabbard/cli/src/main.rs:155]: Like Peter said, I was under the impression that the key used for signing transactions would be the same key used for signing json web tokens

File: services/scabbard/libscabbard/src/client/mod.rs [hunk 11]
```diff
@@ -68,7 +64,15 @@ impl ScabbardClient {
         let body = batches.into_bytes()?;
 
         debug!("Submitting batches via {}", url);
-        let request = Client::new().post(url).body(body);
+        // Allowing unused_mut because request must be mutable if experimental feature
```
Comment (ltseeley) [inline services/scabbard/libscabbard/src/client/mod.rs:69]: These changes to the requests should be in the commit that adds the `auth` field to the client. The cleanest way to break this up would probably be to remove all auth stuff from the builder commit, then add the auth field to the builder, client, and requests in the 2nd commit.

File: services/scabbard/cli/src/main.rs (hunk 10 above)
Comment (ltseeley) [reply services/scabbard/cli/src/main.rs:155]: Didn't realize we were assuming the keys were the same. Sounds good.

File: services/scabbard/libscabbard/src/client/mod.rs (hunk 11 above)
Comment (isabeltomb) [reply services/scabbard/libscabbard/src/client/mod.rs:69]: Updated commits

**Review by ltseeley — APPROVED** [review]

**Review by shannynalayna — APPROVED** [review]
